enso/experiment/rationalized.py

```diff
@@ -306,3 +313,199 @@ def _featurize(self, doc):
+            bsize = batch_end - batch_start
+            doc_vector_slice = doc_vectors[batch_start:batch_end]
+            lengths = list(map(len, doc_vector_slice))
+            max_length = max(lengths)
+            doc_tensor = np.zeros((bsize, max_length, 300))
+            for i, doc_vector in enumerate(doc_vector_slice):
+                doc_tensor[i, :doc_vector.shape[0], :] = doc_vector
+            
+            length_mask = np.zeros((bsize, max_length))
+            for i in range(bsize):
+                length_mask[i, :lengths[i]] = 1.
+
+            yield (doc_tensor, length_mask)
+
+    def featurize_x(self, X):
+        doc_vectors = []
+        for text in X:
+            doc = self.NLP(text, disable=['ner', 'tagger', 'textcat'])
+            word_vectors = np.zeros((len(doc), 300))
+            for i, token in enumerate(doc):
+                word_vectors[i,:] = token.vector
+            doc_vectors.append(word_vectors)
+        return doc_vectors
+
+    def featurize_x_y(self, X, Y):
+        doc_vectors = []
+        rationale_targets = []
+        targets = []
+        for text, (rationales, class_target) in zip(X, Y):
+            doc = self.NLP(text, disable=['ner', 'tagger', 'textcat'])
+            word_vectors = np.zeros((len(doc), 300))
+            rationale_one_hot = np.zeros((len(doc), self.n_classes))
+            target_one_hot = self.target_encoder.transform([class_target])[0]
+            for i, token in enumerate(doc):
+                word_vectors[i,:] = token.vector
+                if self._token_in_rationales(token, rationales):
+                    rationale_one_hot[i,:] = target_one_hot
+
+            doc_vectors.append(word_vectors)
+            rationale_targets.append(rationale_one_hot)
+            targets.append(target_one_hot)
+
+        # (batch, seq, hidden) * (batch, seq, one_hot)
+        stacked_docs = np.vstack(doc_vectors)
+        stacked_targets = np.vstack(rationale_targets)
+
+        rationale_proto = np.einsum('sh,st->ht', stacked_docs, stacked_targets)
+        rationale_proto /= np.einsum('st->t', stacked_targets)
+
+        return doc_vectors, rationale_targets, targets, rationale_proto
+        
+      
+    def fit(self, X, Y):
+        # No rationale means everything is a rationale
+        for x, y in zip(X, Y):
+            if not y[0]:
+                y[0].append({"start": 0, "end": len(x), "label": y[1], "text": x})
+       
+        self.target_encoder.fit([y[1] for y in Y])
+        self.n_classes = len(self.target_encoder.classes_)
+
+        doc_vectors, rationale_targets, targets, proto = self.featurize_x_y(X, Y)
+        train = self.train_iter(doc_vectors, rationale_targets, targets, batch_size=2, n_epochs=500)
+
+        model_fn = lambda x, length_mask: TinyNet(kernel_size=1, n_classes=self.n_classes, rationale_proto=proto)(x, length_mask)
+        model = hk.transform(model_fn)
+        rng = jax.random.PRNGKey(0)
+        sample_data_point = next(train)
```

Comment:
We need a datapoint to infer some shapes on init.  Should just do thi by passing in fake data rather than actually producing an example.